tests/unit/test_in_mem_pub.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2021 - 2022 Universität Tübingen, DKFZ and EMBL
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Testing the in-memory publisher."""
+
+import pytest
+from black import nullcontext
+
+from hexkit.providers.testing.in_mem_pub import InMemEventPublisher, InMemEventStore
+from hexkit.utils import NonAsciiStrError
+
+
+@pytest.mark.parametrize(
+    "type_, key, topic, exception",
+    [
+        ("test_type", "test_key", "test_topic", None),
+        (
+            "test_ßtype",  # non ascii
+            "test_key",
+            "test_topic",
+            NonAsciiStrError,
+        ),
+        (
+            "test_type",
+            "test_ßkey",  # non ascii
+            "test_topic",
+            NonAsciiStrError,
+        ),
+        (
+            "test_type",
+            "test_key",
+            "test_ßtopic",  # non ascii
+            NonAsciiStrError,
+        ),
+    ],
+)
+def test_in_mem_publisher(type_, key, topic, exception):
+    """Test the InMemEventPublisher testing utilities."""
+    payload = {"test_content": "Hello World"}
+
+    # create an in memory event store:
+    event_store = InMemEventStore()
+
+    # create event publisher:
+    event_publisher = InMemEventPublisher(event_store=event_store)
+
+    # publish event using the provider:
+    with (pytest.raises(exception) if exception else nullcontext()):
+        event_publisher.publish(
+            payload=payload,
+            type_=type_,
+            key=key,
+            topic=topic,
+        )
+
+    if not exception:
+        # check if producer was correctly used:
+        stored_event = event_store.get(topic)
+        assert stored_event.key == key
+        assert stored_event.payload == payload
+        assert stored_event.type_ == type_
```

Comment:
```suggestion
        assert stored_event.type_ == type_
        
    with pytest.raises(TopicExhaustedError):
        event_store.get(topic)
```

Add a test for that error.

Reply:
We should add a test with more than one event to make sure it's really a FIFO behavior. The test above with one event would also pass if you had used `pop` instead of `popleft`. And another test to make sure the events are properly assigned to the topics.

Something like this:

```python
def test_in_mem_publisher_with_two_events():
    """Test that events are passed in the right order."""
    payload1 = {"test_content": "Hello"}
    payload2 = {"test_content": "World"}
    type_, key, topic = "test_type", "test_key", "test_topic"
    event_publisher = InMemEventPublisher()
    event_store = event_publisher.event_store
    event_publisher.publish(payload=payload1, type_=type_, key=key, topic=topic)
    event_publisher.publish(payload=payload2, type_=type_, key=key, topic=topic)
    assert event_store.get(topic).payload is payload1
    assert event_store.get(topic).payload is payload2
    with pytest.raises(TopicExhaustedError):
        event_store.get(topic)


def test_in_mem_publisher_with_two_topics():
    """Test that events are passed in the right order."""
    payload1, topic1 = {"test_content": "Hello"}, "test_topic1"
    payload2, topic2 = {"test_content": "World"}, "test_topic2"
    type_, key = "test_type", "test_key"
    event_publisher = InMemEventPublisher()
    event_store = event_publisher.event_store
    event_publisher.publish(payload=payload1, type_=type_, key=key, topic=topic1)
    event_publisher.publish(payload=payload2, type_=type_, key=key, topic=topic2)
    assert event_store.get(topic1).payload is payload1
    with pytest.raises(TopicExhaustedError):
        event_store.get(topic1)
    assert event_store.get(topic2).payload is payload2
    with pytest.raises(TopicExhaustedError):
        event_store.get(topic2)
```

Reply:
Good point.